Port replication_pad1d to ATen

ezyang opened the pull request:
Summary: port replication_pad1d

Differential Revision: D13531920

t-vi commented on aten/src/ATen/native/cuda/ReplicationPadding.cu:
```diff
@@ -0,0 +1,239 @@
+    if (outputPointId >= output.size(2)) {
+    return;
+    }
+    int outputPointX = outputPointId % output.size(2);
```
It's not the problem we're looking for, but Is that a thing or copypasting from 2d? (see the if ... return above).

t-vi commented on aten/src/ATen/native/cuda/ReplicationPadding.cu:
```diff
@@ -0,0 +1,239 @@
+template <typename Dtype>
+  __global__ void replication_pad(
+      const PackedTensorAccessor<Dtype, 3>& input,
+      PackedTensorAccessor<Dtype, 3>& output,
```
I don't think we want to pass the `PackedTensorAccessor`s by reference. To me, the main thing about them (compared to `TensorAccessor` is that they carry their sizes/strides).
I think DeviceTensor worked the same way.

ezyang replied:
Oooh, that's exactly right. I feel embarrassed for missing it!

t-vi replied:
You found so many bugs in my PRs, I'm glad that after so many of them, I get to return the favour. (And I only saw it after debugging in the direction of *are the offsets into the tensors invalid?* for quite a while...)

ezyang commented:
CC @lhuang04 for the comments

lhuang04 committed:
Port replication_pad1d to ATen (#15485)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15485

port replication_pad1d

Differential Revision: D13531920

fbshipit-source-id: 3d107c0314cc19591fc6973a07a964a4341c0db1
